akka23-conductr-client-lib/src/main/scala/com/typesafe/conductr/clientlib/akka/ControlClient.scala

```diff
@@ -63,6 +66,86 @@ class ControlClient(handler: ConnectionHandler, conductrAddress: URL, apiVersion
+   *         - BundleGetSuccess if the get bundle request has been succeeded. This object contains the bundle id, bundle file, and optionally the config file.
+   *         - BundleGetFailure if the get bundle request has been failed. This object contains the HTTP status code and error message.
+   */
+  override def getBundle(bundleId: BundleId)(implicit cc: ConnectionContext): Future[BundleGetResult] =
+    handler.withConnectedRequest(Payload.getBundle(bundleId)) { (responseCode, responseHeader, responseEntity) =>
+      import cc.actorMaterializer
+      import cc.actorMaterializer.executionContext
+
+      def bundleGet: Future[BundleGetResult] = {
+        def writeToDisk(filename: String, source: Source[ByteString, _]): Future[Path] = {
+          val tmpDir = new File(System.getProperty("java.io.tmpdir"))
+          val targetFile = new File(tmpDir, filename)
+          source
+            .runWith(FileIO.toFile(targetFile))
+            .map(_ => targetFile.toPath)
+        }
+
+        def writeBundleFile(bundleParts: Seq[(String, String, Multipart.FormData.BodyPart)]): Future[Path] =
+          bundleParts match {
+            case Seq(entry) if entry._1 == "bundle" =>
+              val (_, fileName, bodyPart) = entry
+              writeToDisk(fileName, bodyPart.entity.dataBytes)
+
+            case _ =>
+              Future.failed(InvalidBundleGetResponseBody("Unable to find bundle file in the response body"))
+          }
+
+        def writeConfigFile(configParts: Seq[(String, String, Multipart.FormData.BodyPart)]): Future[Option[Path]] =
+          configParts.find(_._1 == "configuration") match {
+            case Some((_, fileName, bodyPart)) =>
+              writeToDisk(fileName, bodyPart.entity.dataBytes).map(Some(_))
+
+            case _ =>
+              Future.successful(None)
+          }
+
+        for {
+          formData <- Unmarshal(responseEntity.withSizeLimit(Long.MaxValue)).to[Multipart.FormData]
+
+          bundleAndOthers <- formData
+            .parts
+            .map { bodyPart =>
+              (
+                bodyPart.contentDispositionHeader.flatMap(_.params.get("name")),
+                bodyPart.contentDispositionHeader.flatMap(_.params.get("filename")),
+                bodyPart
+              )
+            }
+            .collect {
+              case (Some(partName), Some(fileName), bodyPart) =>
+                (partName, fileName, bodyPart)
+            }
+            .prefixAndTail(1)
+            .runWith(Sink.head)
+
+          (bundleParts, otherParts) = bundleAndOthers
+          bundleFile <- writeBundleFile(bundleParts)
+
+          configAndOthers <- otherParts.prefixAndTail(1).runWith(Sink.head)
+          (configParts, _) = configAndOthers
+          configFile <- writeConfigFile(configParts)
+        } yield {
+          BundleGetSuccess(bundleId, bundleFile, configFile)
+        }
+      }
```

Comment:
Do you mean `Publisher[Array[Byte]]`? Is there a reason why this is more preferable compared to `Publisher[Byte]`? 

Since a file is a sequence of bytes, I'd naturally think of `Publisher[Byte]` since `Publisher[Byte]` represents sequence of bytes of indeterminate length.

Also using `Publisher[Array[Byte]]` would imply the data will need to be chunked in a certain way, i.e. chunked for every 8kb - am I correct to say this?